lib/editors.js

```diff
@@ -0,0 +1,1477 @@
+                    if (ERR(err, callback)) return;
+                    callback(null);
+                });
+            },
+        ], (err) => {
+            if (ERR(err, callback)) return;
+            callback(null);
+        });
+    }
+}
+
+class QuestionTransferEditor extends Editor {
+    constructor(params) {
+        super(params);
+        this.from_qid = params.from_qid;
+        this.from_course_short_name = params.from_course_short_name;
+        this.from_path = params.from_path;
+        this.description = `Copy question ${this.from_qid} from course ${this.from_course_short_name}`;
+    }
+
+    write(callback) {
+        debug('QuestionTransferEditor: write()');
+        const questionsPath = path.join(this.locals.course.path, 'questions');
+        async.waterfall([
+            (callback) => {
+                debug(`Generate unique QID`);
+                fs.readdir(questionsPath, (err, filenames) => {
+                    if (ERR(err, callback)) return;
+
+                    if (filenames.includes(this.from_qid)) {
+                        let number = 1;
+                        filenames.forEach((filename) => {
+                            let found = filename.match(/^question-([0-9]+)$/);
+                            if (found) {
+                                const foundNumber = parseInt(found[1]);
+                                if (foundNumber >= number) {
+                                    number = foundNumber + 1;
+                                }
+                            }
+                        });
+                        this.qid = `question-${number}`;
+                    } else {
+                        this.qid = this.from_qid;
+                    }
+
+                    this.questionPath = path.join(questionsPath, this.qid);
+                    this.pathsToAdd = [
+                        this.questionPath,
+                    ];
+                    this.commitMessage = `copy question ${this.from_qid} (from course ${this.from_course_short_name}) to ${this.qid}`;
+                    callback(null);
+                });
+            },
+            (callback) => {
+                const fromPath = this.from_path;
+                const toPath = this.questionPath;
+                debug(`Copy template\n from ${fromPath}\n to ${toPath}`);
+                fs.copy(fromPath, toPath, {overwrite: false, errorOnExist: true}, (err) => {
+                    if (ERR(err, callback)) return;
+                    callback(null);
+                });
+            },
+            (callback) => {
+                debug(`Read info.json`);
+                fs.readJson(path.join(this.questionPath, 'info.json'), (err, infoJson) => {
+                    if (ERR(err, callback)) return;
+                    callback(null, infoJson);
+                });
+            },
+            (infoJson, callback) => {
+                debug(`Write info.json with new title and uuid`);
```

Comment:
Don't we need to also change the UUID? I'm thinking about a case where we transfer a question from course A to course B, and then we transfer it back again. If we retain the same UUID we are going to cause a conflict?

Reply:
Yep, it already does change the uuid. 😄